Feature/federate user operations

ivakegg opened the pull request:
This PR is adding the capability to invoke a remote user operations endpoint (i.e. listEffectiveAuthorizations) to enable a client to determine the effective authorizations in total that can be used across the local and remote systems.  The following changes are in here:
1. Added a RemoteUserOperations interface and a RemoteUserOperationsImpl.  These work just like the existing RemoteQueryService and RemoteQueryServiceImpl.
2. Added a remoteUserOperations attribute on the QueryLogic and BaseQueryLogic which is expected to be null for all local query logics
3. The UserOperationsBean will aggregate any configured RemoteUserOperations into its responses by default.
4. The RunningQuery will downgrade user authorizations base on the combined auths of the local and remote systems.
5. The CompositeQueryLogic will subsequent subset the auths when calling the various underlying query logics.
6. Added a ConditionalRemoteUserOperations class that can be used to filter calls to the underlying RemoteUserOperations delegate.  This can be used to, for example, avoid calling the underlying listEffectiveAuthorizations is it appears that the local principal may not need any different auths.
7. Added a trusted dn mechanism that will allow the webserver to remove trusted servers from the proxied entity change.

ivakegg committed:
re #1821: Added a remote user operations implementation
  * Added optional flag to the listEffectiveAuthorizations call to federate
  * Updated Query to now know about a remote user operations
  * Authorization utilities will use the remote user operation if set
  * Composite Query Logic will recalculate query auths for each logic
  * Added test cases for new classes
  * Updated test cases to handle new query logic getUserOperations call
  * Added more of an integration test case which verifies different
  * authorizations on separate query logics

ivakegg commented on warehouse/query-core/src/test/resources/datawave/query/QueryLogicFactory.xml:
```diff
@@ -278,4 +278,7 @@
     
     <bean id="easyRoleManager" class="datawave.webservice.query.logic.EasyRoleManager" />
     
+    <util:list id="RemoteUserServices" value-type="datawave.webservice.common.remote.RemoteUserOperations">
```
Used to configure the UserOperationsBean with any additional remote user operations

ivakegg commented on web-services/client/src/main/java/datawave/webservice/query/QueryImpl.java:
```diff
@@ -413,6 +414,12 @@ public QueryImpl duplicate(String newQueryName) {
         if (null != this.parameters && !this.parameters.isEmpty())
             query.setParameters(new HashSet<Parameter>(this.parameters));
         query.setDnList(this.dnList);
+        if (null != this.optionalQueryParameters && !this.optionalQueryParameters.isEmpty()) {
```
These were oversights in the duplicate method for a QueryImpl

drewfarris replied:
We use `StringUtils` for a null-safe check in many places, why not use `MapUtils` too?
```suggestion
        if (!MapUtils.isEmpty(this.optionalQueryParameters)) {
```

ivakegg commented on web-services/query/src/main/java/datawave/webservice/query/logic/composite/CompositeQueryLogic.java:
```diff
@@ -135,7 +154,7 @@ public void run() {
     protected static final Logger log = Logger.getLogger(CompositeQueryLogic.class);
     
-    private List<QueryLogic<?>> queryLogics = null;
+    private Map<String,QueryLogic<?>> queryLogics = null;
```
Changed to a map for easier relabeling of underlying child query logics.

drewfarris commented on warehouse/query-core/src/main/java/datawave/query/tables/RemoteEventQueryLogic.java:
```diff
@@ -255,4 +258,13 @@ public EventBase transform(EventBase input) throws EmptyObjectException {
+    }
+    
+    @Override
+    public RemoteUserOperations getUserOperations() {
```
Consider renaming the getter to `getRemoteUserOperations` be congruent with the setter
```suggestion
    public RemoteUserOperations getRemoteUserOperations() {
```

ivakegg replied:
sure, ok

drewfarris commented on web-services/common-util/src/main/java/datawave/security/util/AuthorizationsUtil.java:
```diff
@@ -86,8 +99,11 @@ public static Set<Authorizations> getDowngradedAuthorizations(String requestedAu
      */
-    public static LinkedHashSet<Authorizations> getDowngradedAuthorizations(String requestedAuths, DatawavePrincipal principal) {
-        
+    public static LinkedHashSet<Authorizations> getDowngradedAuthorizations(String requestedAuths, DatawavePrincipal principal, RemoteUserService userService)
```
Please update the javadoc for this method to include a description of what role the `RemoteUserService` plays in this process.

ivakegg replied:
ok

drewfarris commented on web-services/common-util/src/main/java/datawave/security/util/AuthorizationsUtil.java:
```diff
@@ -105,11 +121,15 @@ public static LinkedHashSet<Authorizations> getDowngradedAuthorizations(String r
         
+        if (userService != null) {
+            principal = userService.getRemoteUser((DatawavePrincipal) principal);
+        }
```
If I'm reading this correctly, replacing the local principal with the remote principal makes the assumption that we always want the remote auths to replace the local auths wholesale. 

I think this means that we will always expect the remote auths to be a superset of the local auths. How much complexity would it add if we assume it would be possible to the systems to have differerent auths? For example, the local system might have auths the remote system doesn't have any vice versa - yet we want to allow the user to submit a single list of desired auths for both systems.

ivakegg replied:
We always want the remote auths to replace the local auths for a remote query logic.  The composite / federated query logic will combine the local principal with the remote principal to allow that query logic to work correctly.  The two sets of auths do not actually have to overlap at all.

drewfarris commented on web-services/common-util/src/test/java/datawave/security/util/AuthorizationsUtilTest.java:
```diff
@@ -64,34 +65,34 @@ public void testMergeAuthorizations() {
     
     @Test
-    public void testDowngradeAuthorizations() {
+    public void testDowngradeAuthorizations() throws AuthorizationException {
```
Consider adding a few tests that include a mock `RemoteUserService`?

ivakegg replied:
right, the code got covered in other test cases but some specific unit tests would be appropriate here.

drewfarris commented on web-services/common-util/src/main/java/datawave/security/util/AuthorizationsUtil.java:
```diff
@@ -64,14 +67,24 @@ public static Set<Authorizations> mergeAuthorizations(String requestedAuths, Col
         }
     }
     
+    public static interface RemoteUserService {
```
Consider making this a top-level interface as opposed a static nested interface in this class. It seems a little excessive, but generally should avoid confusion in the future. It seems pretty obvious as to what it does, but minimal javadoc could be helpful down the road.

drewfarris replied:
Is this really necessary, considering the fact that `RemoteUserOperations` is an interface? Consider getting rid of either `RemoteUserService` or `RemoteUserOperations`

ivakegg replied:
I guess getting rid of would be reasonable and result in less confusion.

drewfarris commented on web-services/security/src/main/java/datawave/security/user/UserOperationsBean.java:
```diff
@@ -109,10 +124,39 @@ public AuthorizationsListBase listEffectiveAuthorizations() {
             name = datawavePrincipal.getShortName();
             
+            // if we have any remote services configured, merge those authorizations in here
+            if (includeRemoteServices && remoteServices != null && !remoteServices.isEmpty()) {
```
Consider `CollectionUtils`?
```suggestion
            if (includeRemoteServices && !CollectionUtils.isEmpty(remoteServices)) {
```

drewfarris commented on web-services/security/src/main/java/datawave/security/user/UserOperationsBean.java:
```diff
@@ -98,8 +113,8 @@ public class UserOperationsBean {
             "text/html"})
-    public AuthorizationsListBase listEffectiveAuthorizations() {
-        
+    public AuthorizationsListBase listEffectiveAuthorizations(@DefaultValue("true") @QueryParam("includeRemoteServices") boolean includeRemoteServices)
```
javadoc update needed to add both `@param` and `@throws`

drewfarris commented on web-services/security/src/main/java/datawave/security/user/UserOperationsBean.java:
```diff
@@ -137,10 +182,22 @@ public AuthorizationsListBase listEffectiveAuthorizations() {
         log.info("Flushing credentials for " + callerPrincipal + " from the cache.");
+        
+        // if we have any remote services configured, then flush those credentials as well
+        if (includeRemoteServices && remoteServices != null && !remoteServices.isEmpty()) {
```
Another opportunity to use `CollectionUtils.isEmpty(..)`

drewfarris commented on web-services/security/src/main/java/datawave/security/user/UserOperationsBean.java:
```diff
@@ -137,10 +182,22 @@ public AuthorizationsListBase listEffectiveAuthorizations() {
             "application/x-protostuff"})
     @PermitAll
-    public GenericResponse<String> flushCachedCredentials() {
+    public GenericResponse<String> flushCachedCredentials(@DefaultValue("true") @QueryParam("includeRemoteServices") boolean includeRemoteServices) {
```
Another opportunity to improve javadoc.

drewfarris commented on web-services/security/src/main/java/datawave/security/user/UserOperationsBean.java:
```diff
@@ -137,10 +182,22 @@ public AuthorizationsListBase listEffectiveAuthorizations() {
+                try {
+                    remote.flushCachedCredentials(callerPrincipal);
+                } catch (Exception e) {
+                    log.error("Failed to flush user from remote user service", e);
```
Consider how we might alert the caller that the remote operation failed. It would not be fun to be the person calling the cache flush to resolve an issue and having it silently fail on the remote with no indication other than the logs.

ivakegg replied:
Added a message attribute to the response which now get populated.

drewfarris commented on web-services/security/src/test/java/datawave/security/authorization/remote/RemoteUserOperationsImplHttpTest.java:
```diff
@@ -0,0 +1,376 @@
+        remote.setExecutorService(null);
+        remote.setObjectMapperDecorator(new DefaultMapperDecorator());
+        remote.setResponseObjectFactory(new MockResponseObjectFactory());
+        remote.setJsseSecurityDomain(new JSSESecurityDomain() {
```
Consider a static inner class instead of an anonymous class here. It would improve readability. Maybe even a top level class - it might be something that someone else could use?

ivakegg replied:
considering

ivakegg replied:
done

drewfarris commented on web-services/security/src/test/java/datawave/security/authorization/remote/RemoteUserOperationsImplHttpTest.java:
```diff
@@ -0,0 +1,376 @@
+package datawave.security.authorization.remote;
```
Nice Test. I could use something like this. I think we have a couple of these 'Http` tests now, are there common items we can extract from them?

drewfarris commented on web-services/security/src/main/java/datawave/security/authorization/remote/RemoteUserOperationsImpl.java:
```diff
@@ -0,0 +1,279 @@
+        }
+    }
+    
+    @Override
```
The level of delegation we need to implement here makes me sad. Do we have to do this everywhere? Could we do it in a base class.

ivakegg replied:
It makes me sad too.  We are fixing this in the query microservices to be a configuration object instead of a mess of setters.  Can we change this later?

drewfarris commented on web-services/security/src/main/java/datawave/security/authorization/remote/RemoteUserOperationsImpl.java:
```diff
@@ -0,0 +1,279 @@
+        return builder.toString();
+    }
+    
+    public static String getEntities(Object callerObject) {
```
Consider `getProxiedEntities` instead of `getEntities` for clarity.

drewfarris commented on web-services/security/src/main/java/datawave/security/authorization/remote/RemoteUserOperationsImpl.java:
```diff
@@ -0,0 +1,279 @@
+            return callerPrincipal.getProxiedUsers().stream().map(u -> new StringBuilder().append('<').append(u.getDn().subjectDN()).append('>'))
+                            .collect(Collectors.joining());
+        } else {
+            throw new RuntimeException("Cannot use " + callerObject.getClass().getName() + " as a caller object");
```
Consider using checked exceptions for `getEntities` and `getIssuers` or at least something less generic (e.g., `IllegalArgumentException`)

ivakegg replied:
done

drewfarris commented on web-services/security/src/main/java/datawave/security/authorization/remote/RemoteUserOperationsImpl.java:
```diff
@@ -0,0 +1,279 @@
+                log.error("Failed to read entity content.  Trying as a VoidResponse.", ioe);
+                log.error(content);
+                VoidResponse response = voidResponseReader.readValue(content);
+                throw new RuntimeException(response.getMessages().toString());
```
Since this method already throws an `IOEexception`, consider just re-throwing `ioe` here? 

If not, consider adding the `ioe` as the init cause for the `RuntimeException` so we don't lose stack informaton?

drewfarris commented on web-services/security/src/main/java/datawave/security/authorization/remote/ConditionalRemoteUserOperations.java:
```diff
@@ -0,0 +1,114 @@
+
+/**
+ * A conditional remote user operations will only invoke the delegate remote service base on a specified function of the local principal
+ */
```
An example of what you mean by `specified function of the local principal` could be helpful here.

drewfarris commented on web-services/security/src/main/java/datawave/security/authorization/remote/ConditionalRemoteUserOperations.java:
```diff
@@ -0,0 +1,114 @@
+        }
+    }
+    
+    private AuthorizationsListBase getAuthorizationsList(Object callerObject) {
```
A brief description of what this is doing and why might be helpful. It looks like we're just trying to get the auths for each proxied user into the authMap and then copy them them from there to the auths object, but I'm a little confused about the need to filter the authMap by the dn of the primary user.

Could we avoid putting the auths of the primary user in the authMap in the first place?

ivakegg replied:
You are right, we do not need to create this fake response.  An empty one will do.

drewfarris commented on web-services/security/src/main/java/datawave/security/authorization/remote/ConditionalRemoteUserOperations.java:
```diff
@@ -0,0 +1,114 @@
+            AuthorizationsListBase.SubjectIssuerDNPair primaryDn = dn(primaryUser.getDn());
+            auths.setUserAuths(primaryDn.subjectDN, primaryDn.issuerDN, authMap.get(dn(primaryUser.getDn())));
+            authMap.entrySet().stream().filter(e -> !e.getKey().equals(primaryDn))
+                            .forEach(e -> auths.addAuths(e.getKey().subjectDN, e.getKey().issuerDN, e.getValue()));
```
Using `stream` in one case and `forEach` in another (and then mixing them) is a little rough here and at `77` - could we use a streams and collectors instead? 

For readability, good old for loops might be preferred. 

Maybe we could simplify this method in general.

apmoriarty replied:
For a code block like this I also have a preference for readable for-loops

ivakegg replied:
removed code

drewfarris commented on web-services/security/.gitignore:
```diff
@@ -1 +1,2 @@
 /target/
+.keystore
```
Since `.keystore` is test output, perhaps we should put in in the `target` directory.

ivakegg replied:
cleaned this up.

drewfarris commented on web-services/common-util/src/main/java/datawave/security/util/AuthorizationsUtil.java:
```diff
@@ -86,8 +99,11 @@ public static Set<Authorizations> getDowngradedAuthorizations(String requestedAu
+            principal = userService.getRemoteUser((DatawavePrincipal) principal);
+        }
         final DatawaveUser primaryUser = principal.getPrimaryUser();
         UserAuthFunctions uaf = UserAuthFunctions.getInstance();
```
Did you consider overriding the default implementation of `UserAuthFunctions` using the `datawave.user.auth.functions.class` property when thinking about this implementation? It seems to be a natural extension point, but is pretty low-level and might not provide the appropriate hooks to customize the behavior you need.

ivakegg replied:
I did consider that but the whole remote user service seemed more to be couple with remote query logics and hence did not necessarily belong in the authorization service.

drewfarris commented on web-services/common/src/main/java/datawave/webservice/common/remote/RemoteUserOperations.java:
```diff
@@ -0,0 +1,55 @@
+/**
+ * A remote query service is one that can pass calls off to another external user operations endpoint
+ */
+public interface RemoteUserOperations extends AuthorizationsUtil.RemoteUserService {
```
Consider eliminating RemoteUserService or RemoteUserOperations - two levels of interfaces here are potentially overkill.

ivakegg replied:
done

drewfarris commented on web-services/query/src/main/java/datawave/webservice/query/logic/QueryLogic.java:
```diff
@@ -379,4 +380,12 @@ default long getResultLimit(Collection<String> dns) {
+     *
+     * @return A remote user operations interface implementation. Null if NA.
+     */
+    RemoteUserOperations getUserOperations();
```
This may be a bridge too far, but It would be really cool if this always returned an implementation we would use for all user operations. Sometimes that implementation would be something that reaches out to a remote. , otherwise, it would just be for local operations. 

In other words, if we're going to add this to some very central Datawave APIs, it could be really nice if it was somewhat generic and not specifically designed around performing remote user operations.

ivakegg replied:
I made this a little more generic.  I am keeping separate implementations because there are several use cases involved surrounding this concept.

drewfarris commented on web-services/query/src/main/java/datawave/webservice/query/util/LookupUUIDUtil.java:
```diff
@@ -314,18 +324,31 @@ public <T> T createUUIDQueryAndNext(final AbstractUUIDLookupCriteria unvalidated
+                    if (logic.getUserOperations() != null) {
+                        userAuths = AuthorizationsUtil.buildUserAuthorizationString(logic.getUserOperations().getRemoteUser((DatawavePrincipal) principal));
+                    } else {
+                        userAuths = AuthorizationsUtil.buildUserAuthorizationString(principal);
```
Consider creating an updated `buildUserAuthorizationString` that takes the user operations object just like `downgradeUserAuthorizations` does. Ideally we could hide details of things like `getRemoteUser` from most other code in datawave (and remove the conditional that checks for null from this class, because ideally this class shouldn't have to worry about it)

ivakegg replied:
sure, that makes sense.

ivakegg replied:
And this actually finds a few other beans that need to be able to handle the remote user service..... sigh.

apmoriarty replied:
Nice catch

ivakegg replied:
The interface for these has now changed.  Please reevaluate whether I got to a better place.

drewfarris commented on web-services/query/src/main/java/datawave/webservice/query/util/LookupUUIDUtil.java:
```diff
@@ -363,12 +386,11 @@ public <T> T createUUIDQueryAndNext(final AbstractUUIDLookupCriteria unvalidated
             // Otherwise, execute the query for a standard paged response
             else {
-                response = (T) this.queryExecutor.createQueryAndNext(queryParameters.getFirst(QueryParameters.QUERY_LOGIC_NAME), queryParameters);
+                response = (T) this.queryExecutor.createQueryAndNext(logicName, queryParameters);
```
This is much nicer than it was before.

drewfarris commented on web-services/query/src/main/java/datawave/webservice/query/logic/BaseQueryLogic.java:
```diff
@@ -395,4 +396,10 @@ public ResponseEnricherBuilder getResponseEnricherBuilder() {
     }
+    
+    @Override
+    public RemoteUserOperations getUserOperations() {
```
Consider removing the distinction between remote and local operations and just providing a `UserOperations` class that gets plugged into the query logics. Different implementations of this would perform either local-only or local-remote operations. This might clean up the API a bit.

ivakegg replied:
Done.

ivakegg replied:
done

drewfarris commented on web-services/examples/client-login/src/main/java/datawave/webservice/examples/ClientLoginExampleBean.java:
```diff
@@ -75,7 +75,7 @@ public void handle(Callback[] callbacks) {
             // Call secured EJBs
             try {
-                AuthorizationsListBase auths = userOps.listEffectiveAuthorizations();
+                AuthorizationsListBase auths = userOps.listEffectiveAuthorizations(false);
```
Consider whether there is a way we can avoid the need to add this flag in the API.

ivakegg replied:
The flag is a query parameter and is optional in the http API.  I can add convience methods to the bean if you think that is better.  I think passing in an explicit flag works better from a maintainance perspective.  What you want thec call to do is more obvious.

drewfarris commented on web-services/query/src/main/java/datawave/webservice/query/logic/composite/CompositeUserOperations.java:
```diff
@@ -0,0 +1,105 @@
+    }
+    
+    @Override
+    public AuthorizationsListBase listEffectiveAuthorizations(Object callerObject) throws AuthorizationException {
```
I feel like we have similar code elsewhere in this PR, I wonder if there are ways we can consolidate?

ivakegg replied:
I think this is the only place where we actually merge authorization, so this is actually unique code.

ivakegg replied:
I did move some of the common logic into the authorizations util however to handle merging users and principals.

drewfarris commented on web-services/query/src/main/java/datawave/webservice/query/logic/composite/CompositeQueryLogic.java:
```diff
@@ -145,52 +164,80 @@ public void run() {
     }
     
+    public static Set<Authorizations> updateRuntimeAuthorizationsAndQueryAuths(QueryLogic<?> logic, Query settings) throws AuthorizationException {
+        Set<String> requestedAuths = new HashSet<>(AuthorizationsUtil.splitAuths(settings.getQueryAuthorizations()));
```
Consider whether it would be better to avoid calling `splitAuths` with each call to this method.

ivakegg replied:
This method only gets called once per create or getPlan call, so I think this is good as it stands.

drewfarris commented on web-services/query/src/main/java/datawave/webservice/query/logic/composite/CompositeQueryLogic.java:
```diff
@@ -145,52 +164,80 @@ public void run() {
+        // determine the valid authorizations for this call to be the user's auths for this logic
+        DatawavePrincipal principal = (DatawavePrincipal) logic.getPrincipal();
+        if (logic.getUserOperations() != null) {
+            principal = logic.getUserOperations().getRemoteUser(principal);
```
See question elsewhere regarding the implications of replacing the local principal with the remote principal.

ivakegg committed:
review comments

ivakegg committed:
Refactored RemoteUserOperations to be common-util

ivakegg committed:
Added auth util unit tests for remote user service

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/tables/RemoteEventQueryLogic.java:
```diff
@@ -255,4 +258,13 @@ public EventBase transform(EventBase input) throws EmptyObjectException {
     }
     
+    @Override
+    public void setRemoteUserOperations(RemoteUserOperations userService) {
```
rename `userService` to either `userOperations` or `remoteUserOperations`

ivakegg replied:
done

apmoriarty commented on warehouse/query-core/src/main/java/datawave/query/tables/remote/RemoteQueryLogic.java:
```diff
@@ -1,11 +1,14 @@
 public interface RemoteQueryLogic<T> extends QueryLogic<T> {
     public void setRemoteQueryService(RemoteQueryService service);
+    
+    public void setRemoteUserOperations(RemoteUserOperations userService);
```
rename `userService` to either `remoteUserOperations` or `remoteOperations`

ivakegg replied:
done

apmoriarty commented on web-services/common-util/src/main/java/datawave/security/util/AuthorizationsUtil.java:
```diff
@@ -64,9 +68,32 @@ public static Set<Authorizations> mergeAuthorizations(String requestedAuths, Col
+     *            principal will be used. Otherwise this service will be used to retrieve the remote principal from which the auths will be determined.
+     * @return A set of {@link Authorizations}, one per entity represented in {@code principal}. The user's auths are replaced by {@code requestedAuths} so long
+     *         as the user actually had all of the auths. If {@code requestedAuths} is {@code null}, then the user's auths are returned as-is.
+     */
```
Good docs

ivakegg replied:
thanks

apmoriarty commented on web-services/pom.xml:
```diff
@@ -40,7 +40,7 @@
         <version.asm>4.0</version.asm>
         <version.axiom>1.2.14</version.axiom>
-        <version.bcprov>1.46</version.bcprov>
+        <version.bcprov>1.67</version.bcprov>
```
Can you provide a quick explanation for this version bump?

ivakegg replied:
Actually this bump messed up some integration, so reworking

apmoriarty commented on web-services/query/src/main/java/datawave/webservice/query/logic/BaseQueryLogic.java:
```diff
@@ -395,4 +396,10 @@ public ResponseEnricherBuilder getResponseEnricherBuilder() {
+    
+    @Override
+    public RemoteUserOperations getRemoteUserOperations() {
+        // null implies that the local user operations/principal is to be used for auths.
```
+1

apmoriarty commented on web-services/query/src/main/java/datawave/webservice/query/logic/composite/CompositeLogicException.java:
```diff
@@ -1,20 +1,17 @@
+    public CompositeLogicException(String message, Map<String,Exception> exceptions) {
         super(getMessage(message, exceptions));
-        exceptions.stream().forEach(e -> addSuppressed(e.getValue()));
+        exceptions.values().stream().forEach(e -> addSuppressed(e));
     }
     
-    private static String getMessage(String message, List<Map.Entry<QueryLogic<?>,Exception>> exceptions) {
+    private static String getMessage(String message, Map<String,Exception> exceptions) {
         StringBuilder builder = new StringBuilder();
         builder.append(message).append(":");
-        exceptions.stream().forEach(e -> builder.append('\n').append(e.getKey().getLogicName()).append(": ").append(e.getValue().getMessage()));
+        exceptions.entrySet().stream().forEach(e -> builder.append('\n').append(e.getKey()).append(": ").append(e.getValue().getMessage()));
```
Nice cleanup. This version is much easier to read.

apmoriarty commented on web-services/query/src/main/java/datawave/webservice/query/logic/composite/CompositeUserOperations.java:
```diff
@@ -0,0 +1,104 @@
+import java.util.Map;
+import java.util.Set;
+
+public class CompositeUserOperations implements RemoteUserOperations {
```
Add a brief description, even if it's only a couple of sentences.

ivakegg replied:
done

apmoriarty commented on web-services/security/pom.xml:
```diff
@@ -82,7 +82,7 @@
         <dependency>
             <groupId>org.bouncycastle</groupId>
-            <artifactId>bcprov-jdk16</artifactId>
+            <artifactId>bcprov-jdk15on</artifactId>
```
Was this a downgrade? Why was this change made?

ivakegg replied:
Reverted this change.

ivakegg committed:
Updated other beans to make use of the remote user operations.

ivakegg committed:
Update web-services/security/src/main/java/datawave/security/user/UserOperationsBean.java

Co-authored-by: Drew Farris <drew.farris@gmail.com>

ivakegg committed:
use collections util

ivakegg committed:
additional review updates

billoley commented on web-services/common-util/src/main/java/datawave/security/authorization/RemoteUserOperations.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.stream.Collectors;
+
+/**
+ * A remote query service is one that can pass calls off to another external user operations endpoint
```
```suggestion
 * A remote user operations service is one that can pass calls off to another external user operations endpoint
```

ivakegg replied:
+1

billoley commented on warehouse/query-core/src/test/resources/datawave/query/QueryLogicFactory.xml:
```diff
@@ -278,4 +278,7 @@
     
     <bean id="easyRoleManager" class="datawave.webservice.query.logic.EasyRoleManager" />
     
+    <util:list id="RemoteUserServices" value-type="datawave.security.authorization.RemoteUserOperations">
```
```suggestion
    <util:list id="RemoteUserOperationsList" value-type="datawave.security.authorization.RemoteUserOperations">
```

ivakegg replied:
+1

billoley commented on web-services/common-util/src/main/java/datawave/security/util/AuthorizationsUtil.java:
```diff
@@ -64,9 +66,32 @@ public static Set<Authorizations> mergeAuthorizations(String requestedAuths, Col
+                    throws AuthorizationException {
         if (principal instanceof DatawavePrincipal) {
-            return getDowngradedAuthorizations(requestedAuths, (DatawavePrincipal) principal);
+            return getDowngradedAuthorizations(requestedAuths, (DatawavePrincipal) principal, userService);
```
```suggestion
            return getDowngradedAuthorizations(requestedAuths, (DatawavePrincipal) principal, remoteUserOperations);
```

billoley commented on web-services/common-util/src/main/java/datawave/security/util/AuthorizationsUtil.java:
```diff
@@ -64,9 +66,32 @@ public static Set<Authorizations> mergeAuthorizations(String requestedAuths, Col
+     * @return A set of {@link Authorizations}, one per entity represented in {@code principal}. The user's auths are replaced by {@code requestedAuths} so long
+     *         as the user actually had all of the auths. If {@code requestedAuths} is {@code null}, then the user's auths are returned as-is.
+     */
+    public static Set<Authorizations> getDowngradedAuthorizations(String requestedAuths, Principal principal, RemoteUserOperations userService)
```
```suggestion
    public static Set<Authorizations> getDowngradedAuthorizations(String requestedAuths, Principal principal, RemoteUserOperations remoteUserOperations)
```

billoley commented on web-services/common-util/src/main/java/datawave/security/util/AuthorizationsUtil.java:
```diff
@@ -83,11 +108,17 @@ public static Set<Authorizations> getDowngradedAuthorizations(String requestedAu
+    public static LinkedHashSet<Authorizations> getDowngradedAuthorizations(String requestedAuths, DatawavePrincipal principal, RemoteUserOperations userService)
+                    throws AuthorizationException {
+        if (userService != null) {
+            principal = userService.getRemoteUser((DatawavePrincipal) principal);
```
```suggestion
            principal = userService.getRemoteUser(principal);
```
principal already is a DatawavePrincipal

billoley commented on web-services/common-util/src/main/java/datawave/security/util/AuthorizationsUtil.java:
```diff
@@ -103,16 +134,19 @@ public static LinkedHashSet<Authorizations> getDowngradedAuthorizations(String r
      * @return requested, unless the user represented by {@code principal} does not have one or more of the auths in {@code requested}
      */
-    public static String downgradeUserAuths(Principal principal, String requested) {
+    public static String downgradeUserAuths(Principal principal, String requested, RemoteUserOperations userService) throws AuthorizationException {
```
```suggestion
    public static String downgradeUserAuths(Principal principal, String requested, RemoteUserOperations remoteUserOperations) throws AuthorizationException {
```

billoley commented on web-services/common-util/src/main/java/datawave/security/util/AuthorizationsUtil.java:
```diff
@@ -103,16 +134,19 @@ public static LinkedHashSet<Authorizations> getDowngradedAuthorizations(String r
         List<String> requestedList = AuthorizationsUtil.splitAuths(requested);
         // Find all authorizations the user has access to
-        String userAuths = AuthorizationsUtil.buildUserAuthorizationString(principal);
+        String userAuths = AuthorizationsUtil.buildUserAuthorizationString(principal, userService);
```
```suggestion
        String userAuths = AuthorizationsUtil.buildUserAuthorizationString(principal, remoteUserOperations);
```

billoley commented on web-services/common-util/src/main/java/datawave/security/util/AuthorizationsUtil.java:
```diff
@@ -166,7 +200,20 @@ public static String buildAuthorizationString(Collection<? extends Collection<St
+     *            principal will be used. Otherwise this service will be used to retrieve the remote principal from which the auths will be determined.
+     * @return user authorizations string
+     */
+    public static String buildUserAuthorizationString(Principal principal, RemoteUserOperations userService) throws AuthorizationException {
```
```suggestion
    public static String buildUserAuthorizationString(Principal principal, RemoteUserOperations remoteUserOperations) throws AuthorizationException {
```

billoley commented on web-services/common-util/src/main/java/datawave/security/util/AuthorizationsUtil.java:
```diff
@@ -166,7 +200,20 @@ public static String buildAuthorizationString(Collection<? extends Collection<St
+    public static String buildUserAuthorizationString(Principal principal, RemoteUserOperations userService) throws AuthorizationException {
+        if (userService != null) {
+            principal = userService.getRemoteUser((DatawavePrincipal) principal);
+        }
```
```suggestion
        if (remoteUserOperations != null) {
            principal = remoteUserOperations.getRemoteUser((DatawavePrincipal) principal);
        }
```

ivakegg committed:
Renamed RemoteUserOperations to UserOperations

ivakegg committed:
* Updated the authorization mechanisms to validate against the overall
user (which may include the remote authorizations) and then the reduce
by the query authorizations.  This allows the auths returned by the list
effective authorizations which include remote auths to be used for local
query logics as well.

ivakegg committed:
Moved a bunch of the logic into utility functions and refined based on
integration testing.

ivakegg committed:
More tests and fixes

ivakegg committed:
Initial attempt at being able to configure a pruning of proxied entities

ivakegg committed:
Updated property for trusted proxied entities

ivakegg committed:
formatting

ivakegg committed:
Fixed import and corrected filter

ivakegg committed:
Automatically add the server dn as a trusted entity.

ivakegg committed:
Externalized class

ivakegg committed:
Avoid some runtime exceptions and use authorization exceptions instead

ivakegg committed:
Refactored a mess of methods back into the base RemoteHttpService class
to simplify the various extended classes.

ivakegg committed:
Removed faking out of the effective auths call

ivakegg committed:
Factored out a test jsse security domain

ivakegg committed:
rename parameter

ivakegg committed:
rename parameter

ivakegg committed:
Added comments

ivakegg committed:
Updated variable names

ivakegg committed:
Updated test to work with bouncy castle 16

ivakegg committed:
javadoc

ivakegg committed:
Unused throws

apmoriarty commented on web-services/common-util/src/main/java/datawave/security/util/AuthorizationsUtil.java:
```diff
@@ -64,12 +71,9 @@ public static Set<Authorizations> mergeAuthorizations(String requestedAuths, Col
-        } else {
-            return Collections.singleton(new Authorizations());
-        }
+    public static DatawaveUser mergeAuths(DatawaveUser user1, Set<String> auths) {
```
`user1` implies more than one user, just call it `user`

ivakegg replied:
ok, will do

apmoriarty commented on web-services/query/src/test/java/datawave/webservice/query/logic/composite/CompositeQueryLogicTest.java:
```diff
@@ -372,9 +442,9 @@ public void setup() {
-        logics.add(new TestQueryLogic());
-        logics.add(new TestQueryLogic());
+        Map<String,QueryLogic<?>> logics = new HashMap<>();
+        logics.put("TestQueryLogic", new TestQueryLogic());
+        logics.put("TestQueryLogic2", new TestQueryLogic());
```
These names should probably be static constants

ivakegg replied:
no need as this is a test case.

apmoriarty commented on web-services/security/src/main/java/datawave/security/authorization/remote/RemoteUserOperationsImpl.java:
```diff
@@ -0,0 +1,83 @@
+        final String suffix = FLUSH_CREDS;
+        // includeRemoteServices=false to avoid any loops
+        return executeGetMethodWithRuntimeException(suffix, uriBuilder -> {
+            uriBuilder.addParameter("includeRemoteServices", "false");
```
`includeRemoteServices` should probably be a static value, similar to `FLUSH_CREDS` and `LIST_EFFECTIVE_AUTHS`

ivakegg replied:
agreed

apmoriarty reviewed (dismissed):
Several cosmetic/stylistic comments. Otherwise looks good.

billoley commented on web-services/common-util/src/main/java/datawave/security/util/AuthorizationsUtil.java:
```diff
@@ -81,57 +85,99 @@ public static Set<Authorizations> getDowngradedAuthorizations(String requestedAu
+        
+        // validate that the query user is actually a subset of the primary user
+        if (!primaryUser.getAuths().containsAll(queryUser.getAuths())) {
+            throw new IllegalArgumentException("The user's query auths are not a subset of the user's overall auths.");
```
"The user's query auths are not a subset of the user's overall auths." seems a little too "inside baseball" with the "overall auths" language --- can we just say that the user requested authorizations that they do not have?

ivakegg replied:
hmm... good point.... let me try and reword that.

billoley commented on web-services/common-util/src/main/java/datawave/security/util/AuthorizationsUtil.java:
```diff
@@ -183,4 +236,72 @@ public static Collection<Authorizations> minimize(Collection<Authorizations> aut
+                    if (users.containsKey(user.getDn())) {
+                        users.put(user.getDn(), mergeUsers(users.get(user.getDn()), user));
+                    } else {
+                        extraProxies.put(user.getDn(), user);
```
Not really maintaining order of the proxiedUsers if any extra entities just get added at the end of the chain.

ivakegg replied:
yea, I was not too sure how to mix that in.  OK, I will try to maintain the order a little better.  good catch.

ivakegg replied:
Actually this is not a concern as the only real possibility is that we have extra proxies at the end of the list.  And even if we do have a different ordering of things (a proxy inserted in the middle), then this really does not matter as long as the primary user is at the front.

ivakegg committed:
review nits

ivakegg committed:
reworded exception message

FineAndDandy commented on web-services/common-util/src/main/java/datawave/security/util/AuthorizationsUtil.java:
```diff
@@ -81,57 +85,99 @@ public static Set<Authorizations> getDowngradedAuthorizations(String requestedAu
+        uaf.validateRequestedAuthorizations(requestedAuths, primaryUser);
+        
+        // now return auths that are a reduced by what the query can handle.
+        return uaf.getRequestedAuthorizations(requestedAuths, queryUser, false).toString();
```
consider merging the duplicate code from this method and the previous one?

ivakegg replied:
good idea.

FineAndDandy commented on web-services/deploy/configuration/src/main/resources/datawave/query/QueryLogicFactory.xml:
```diff
@@ -132,11 +132,25 @@
+    </bean>
+
+    <!-- Configure the remote operation here for use by the local UserOperationsBean -->
+    <util:list id="RemoteUserOperationsList" value-type="datawave.security.authorization.RemoteUserOperations">
+        <!--ref bean="MyRemoteUserOps"/-->
+    </util:list>
```
I can't tell if this was commented out and forgotten or intended to be an illustration

ivakegg replied:
intended illustration

FineAndDandy commented on web-services/query/src/main/java/datawave/webservice/query/logic/QueryLogic.java:
```diff
@@ -379,4 +380,12 @@ default long getResultLimit(Collection<String> dns) {
+     *
+     * @return A user operations interface implementation. Null if NA (i.e. the local principal is sufficient)
+     */
+    UserOperations getUserOperations();
```
How would you feel about renaming this to remoteUserOperations()? The name feels misleading. If there are no remote user operations then the local principal is sufficient.

ivakegg replied:
That is funny, Drew had me rename it from remote user operations to this.

ivakegg replied:
I am leaving it as is.  maybe someday this would be non-null and yet may be local somehow.

FineAndDandy commented on web-services/query/src/main/java/datawave/webservice/query/logic/composite/CompositeUserOperations.java:
```diff
@@ -0,0 +1,125 @@
+            auths.setUserAuths(primaryDn.subjectDN, primaryDn.issuerDN, authMap.get(dn(primaryUser.getDn())));
+            authMap.entrySet().stream().filter(e -> !e.getKey().equals(primaryDn))
+                            .forEach(e -> auths.addAuths(e.getKey().subjectDN, e.getKey().issuerDN, e.getValue()));
+        }
```
should there be a warning/error/exception thrown if callerObject is not a DatawavePrincipal?

ivakegg replied:
+1

FineAndDandy commented on web-services/query/src/main/java/datawave/webservice/query/logic/composite/CompositeUserOperations.java:
```diff
@@ -0,0 +1,125 @@
+            GenericResponse<String> remoteResponse = ops.flushCachedCredentials(callerObject);
+            if (remoteResponse.getHasResults()) {
+                response.setHasResults(true);
+                response.setResult(response.getResult() + separator + remoteResponse.getResult());
+                separator = "\n";
```
It isn't clear to me what the intended use of separator is here.

ivakegg replied:
To separate

FineAndDandy commented on web-services/query/src/main/java/datawave/webservice/query/remote/RemoteQueryServiceImpl.java:
```diff
@@ -92,6 +74,11 @@ public GenericResponse<String> planQuery(String queryLogicName, Map<String,List<
     }
     
     private GenericResponse<String> query(String endPoint, String queryLogicName, Map<String,List<String>> queryParameters, Object callerObject) {
+        if (!(callerObject instanceof DatawavePrincipal)) {
+            throw new RuntimeException("Cannot handle a " + callerObject.getClass() + ". Only DatawavePrincipal is accepted");
+        }
+        final DatawavePrincipal principal = (DatawavePrincipal) callerObject;
```
consider private DatawavePrincipal getPrincipal(Object callerObject)

ivakegg replied:
+1

FineAndDandy commented on web-services/security/src/main/java/datawave/security/authorization/remote/ConditionalRemoteUserOperations.java:
```diff
@@ -0,0 +1,85 @@
+public class ConditionalRemoteUserOperations implements UserOperations {
+    private static final Logger log = LoggerFactory.getLogger(ConditionalRemoteUserOperations.class);
+    
+    private UserOperations delegate;
+    private Function<DatawavePrincipal,Boolean> condition;
+    private ResponseObjectFactory responseObjectFactory;
```
None of these objects are null checked in this class

ivakegg replied:
+1

FineAndDandy commented on web-services/security/src/main/java/datawave/security/authorization/remote/RemoteUserOperationsImpl.java:
```diff
@@ -0,0 +1,84 @@
+    
+    @Override
+    public AuthorizationsListBase listEffectiveAuthorizations(Object callerObject) throws AuthorizationException {
+        if (!(callerObject instanceof DatawavePrincipal)) {
```
check initialized?

ivakegg replied:
Good catch!

FineAndDandy commented on web-services/security/src/main/java/datawave/security/authorization/remote/RemoteUserOperationsImpl.java:
```diff
@@ -0,0 +1,84 @@
+    
+    @Override
+    public GenericResponse<String> flushCachedCredentials(Object callerObject) throws AuthorizationException {
+        if (!(callerObject instanceof DatawavePrincipal)) {
+            throw new AuthorizationException("Cannot handle a " + callerObject.getClass() + ". Only DatawavePrincipal is accepted");
+        }
+        final DatawavePrincipal principal = (DatawavePrincipal) callerObject;
+        final String suffix = FLUSH_CREDS;
```
check initialized?

ivakegg replied:
Good catch!

FineAndDandy reviewed (changes requested):
I found myself having to keep going back to review principal, queryPrincipal, remotePrincipal, and  how they all worked together to create auths. You have comments scattered around, but maybe a readme or extended class javadoc to put all the information in one place could be helpful for future maintenance
